File: Arrays/Plus One.cpp
class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        int lsb = digits.size() - 1, carry = 0;
        while(lsb >= 0){
            if(digits[lsb] < 9){
                digits[lsb]++;
                break;
            }
            lsb--;
        }
        if(lsb == -1){
            digits.insert(digits.begin(), 1);
            lsb = 0;
        }
        lsb = lsb+1;
        while(lsb < digits.size() && digits[lsb] == 9){
            digits[lsb] = 0;
            lsb++;
        }
        return digits;
    }
};

File: Arrays/Plus One.rs
impl Solution {
    pub fn plus_one(mut digits: Vec<i32>) -> Vec<i32> {
        let n = digits.len();
        let mut carry = false;

        if digits[n-1] == 9 {
            digits[n-1] = 0;
            carry = true;
        } else {
            digits[n-1] += 1;
        }

        for i in (0..n-1).rev() {
            if !carry {
                break;
            }

            if digits[i] == 9 {
                digits[i] = 0;
            } else {
                digits[i] += 1;
                carry = false;
            }
        }

        if carry {
            digits.insert(0, 1);
        }
        digits
    }
}